pyomo/solvers/plugins/solvers/SAS.py

```diff
@@ -0,0 +1,701 @@
+                primal_out = primal_out.set_index('_VAR_', drop=True)
+                primal_out = primal_out.rename({'_VALUE_': 'Value'}, axis='columns')
+                sol.variable = primal_out.to_dict('index')
+
+        self._rc = 0
+        return Bunch(rc=self._rc, log=self._log)
+
+
+class SASLogWriter:
+    """Helper class to take the log from stdout and put it also in a StringIO."""
+
+    def __init__(self, tee):
+        """Set up the two outputs."""
+        self.tee = tee
+        self._log = StringIO()
+        self.stdout = sys.stdout
+
+    def write(self, message):
+        """If the tee options is specified, write to both outputs."""
+        if self.tee:
+            self.stdout.write(message)
+        self._log.write(message)
+
+    def flush(self):
+        """Nothing to do, just here for compatibility reasons."""
+        # Do nothing since we flush right away
+        pass
+
+    def log(self):
+        """ "Get the log as a string."""
+        return self._log.getvalue()
+
+
+@SolverFactory.register('_sascas', doc='SAS Viya CAS Server interface')
+class SASCAS(SASAbc):
+    """
+    Solver interface connection to a SAS Viya CAS server using swat.
+    See the documentation for the swat package about how to create a connection.
+    The swat connection options can be passed as options to the solve function.
+    """
+
+    def __init__(self, **kwds):
+        """Initialize and try to load the swat package."""
+        super(SASCAS, self).__init__(**kwds)
+
+        try:
+            import swat
+
+            self._sas = swat
+        except ImportError:
+            self._python_api_exists = False
+        except Exception as e:
+            self._python_api_exists = False
+            # For other exceptions, raise it so that it does not get lost
+            raise e
+        else:
+            self._python_api_exists = True
+
+    def _apply_solver(self):
+        """ "Prepare the options and run the solver. Then store the data to be returned."""
+        logger.debug("Running SAS Viya")
+
+        # Set return code to issue an error if we get interrupted
+        self._rc = -1
+
+        # Extract CAS connection options
+        cas_opts = {}
+        for opt in CAS_OPTION_NAMES:
+            val = self.options.pop(opt, None)
+            if val != None:
+                cas_opts[opt] = val
+
+        # Figure out if the problem has integer variables
+        with_opt = self.options.pop("with", None)
+        if with_opt == "lp":
+            action = "solveLp"
+        elif with_opt == "milp":
+            action = "solveMilp"
+        else:
+            # Check if there are integer variables, this might be slow
+            action = "solveMilp" if self._has_integer_variables() else "solveLp"
+
+        # Connect to CAS server
+        with redirect_stdout(SASLogWriter(self._tee)) as self._log_writer:
+            s = self._sas.CAS(**cas_opts)
+            try:
```

Comment:
Is it possible to break this up into smaller, helper functions? It's quite large as-is, and I find it difficult to track all of the logic.